exporter/otlpexporter/arrow.go

```diff
@@ -0,0 +1,469 @@
+	ae.cancel()
+	ae.wg.Wait()
+	return nil
+}
+
+// setBatchChannel places a waiting consumer's batchID into the waiters map, where
+// the stream reader may find it.
+func (stream *arrowStream) setBatchChannel(batchID string, errCh chan error) {
+	stream.lock.Lock()
+	defer stream.lock.Unlock()
+
+	stream.waiters[batchID] = errCh
+}
+
+// writeStream repeatedly places this stream into the next-available queue, then
+// performs a blocking send().  This returns when the data is in the write buffer,
+// the caller waiting on its error channel.
+func (ae *arrowExporter) writeStream(ctx context.Context, stream *arrowStream, ww *sync.WaitGroup) {
+	defer ww.Done()
+	defer stream.cancel()
+
+writeLoop:
+	for {
+		// Note: this can't block b/c stream has capacity &
+		// individual streams shut down synchronously.
+		ae.ready.setReady(stream)
+
+		// this can block, and if the context is canceled we
+		// wait for the reader to find this stream.
+		var wri writeItem
+		select {
+		case wri = <-stream.toWrite:
+		case <-ctx.Done():
+			// Because we did not <-stream.toWrite, there is a potential
+			// sender race happening, which is handled below.
+			break writeLoop
+		}
+		// Note: For the two return statements below there is no potential
+		// sender race because the stream is not available, as indicated by
+		// the successful <-stream.toWrite.
+
+		batch, err := stream.encode(wri.records)
+		if err != nil {
+			// TODO: Is this not permanent?  Another
+			// sequence of data might not produce it.
+			//
+			// This is some kind of internal error.
+			wri.errCh <- consumererror.NewPermanent(err)
+			ae.exporter.settings.Logger.Error("arrow encode", zap.Error(err))
+			return
+		}
+
+		stream.setBatchChannel(batch.BatchId, wri.errCh)
+
+		if err := stream.client.Send(batch); err != nil {
+			// The error will be sent to errCh during cleanup for this stream.
+			ae.exporter.settings.Logger.Error("arrow send", zap.Error(err))
+			return
+		}
+	}
+
+	// Remove this from the ready channel.
+	ae.ready.removeReady(stream)
+}
+
+// readStream repeatedly reads a batch status and releases the consumers waiting for
+// a response.
+func (ae *arrowExporter) readStream(ctx context.Context, stream *arrowStream) error {
+	defer stream.cancel()
+
+	for {
+		resp, err := stream.client.Recv()
+		if err != nil {
+			return err
+		}
+
+		if err = stream.processBatchStatus(resp.Statuses); err != nil {
+			return err
+		}
+	}
+}
+
+// getSenderChannels takes the stream lock and removes the corresonding sender
+// channel for each BatchId.
+func (stream *arrowStream) getSenderChannels(statuses []*arrowpb.StatusMessage) ([]chan error, error) {
+	var err error
+
+	fin := make([]chan error, len(statuses))
+
+	stream.lock.Lock()
+	defer stream.lock.Unlock()
+
+	for idx, status := range statuses {
+		ch, ok := stream.waiters[status.BatchId]
+		if !ok {
+			// Will break the stream.
+			err = multierr.Append(err, fmt.Errorf("duplicate stream response: %s", status.BatchId))
```

Comment:
Is it also possible for the batchId to not be in the waiters map? I think the answer is no, but good to check that expectation.

Reply:
This does have the potential issue for having nils in the returned fin map ([playground](https://go.dev/play/p/gyIuO4AaYO9)) which means in processBatchStatus we may attempt to write to a nil channel

Reply:
This is fairly defensive code. If gRPC is being correct about response order, we should't see duplicates, so this appears to be a confused receiver.

Part 2: Good catch. I forgot to check for `nil` when scanning the returned list.